authorization/v2/uniformResourceIdentifier_test.go

```diff
@@ -42,3 +42,180 @@ func TestNewUriFromRequest(t *testing.T) {
 	}, uri.arguments)
 	assert.Nil(t, uri.metadata)
 }
+
+func Test_NewURIFromString__should_return_correct_URI(t *testing.T) {
+	tests := []struct {
+		name        string
+		uri         string
+		expectedURI UniformResourceIdentifier
+	}{
+		{
+			name: "with only path",
+			uri:  "hs:hs_auth:api:v2:provide_access_to_uri",
+			expectedURI: UniformResourceIdentifier{
+				path:      "hs:hs_auth:api:v2:provide_access_to_uri",
+				arguments: nil,
+				metadata:  nil,
+			},
+		},
+		{
+			name: "with path and arguments",
+			uri:  "hs:hs_auth:api:v2:provide_access_to_uri?allowed_uri%3Dhs%3Ahs_application%3A%2A",
+			expectedURI: UniformResourceIdentifier{
+				path:      "hs:hs_auth:api:v2:provide_access_to_uri",
+				arguments: map[string]string{"allowed_uri": "hs:hs_application:*"},
+				metadata:  nil,
+			},
+		},
+		{
+			name: "with path, arguments and metadata",
+			uri:  "hs:hs_auth:api:v2:provide_access_to_uri?allowed_uri%3Dhs%3Ahs_application%3A%2A#until%3D21392103",
+			expectedURI: UniformResourceIdentifier{
+				path:      "hs:hs_auth:api:v2:provide_access_to_uri",
+				arguments: map[string]string{"allowed_uri": "hs:hs_application:*"},
+				metadata:  map[string]string{"until": "21392103"},
+			},
+		},
+		{
+			name: "with path and metadata",
+			uri:  "hs:hs_auth:api:v2:provide_access_to_uri#until%3D21392103",
+			expectedURI: UniformResourceIdentifier{
+				path:      "hs:hs_auth:api:v2:provide_access_to_uri",
+				arguments: nil,
+				metadata:  map[string]string{"until": "21392103"},
+			},
+		},
+		{
+			name: "with url encoded metadata rune in metadata",
+			uri:  "hs:hs_auth:api:v2:provide_access_to_uri?test=ok%23#test2=ok",
+			expectedURI: UniformResourceIdentifier{
+				path:      "hs:hs_auth:api:v2:provide_access_to_uri",
+				arguments: map[string]string{"test": "ok#"},
+				metadata:  map[string]string{"test2": "ok"},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			actualURI, err := NewURIFromString(tt.uri)
+			assert.NoError(t, err)
+
+			assert.Equal(t, actualURI, tt.expectedURI)
+		})
+	}
+}
+
+func Test_NewURIFromString__should_throw_error(t *testing.T) {
+	tests := []struct {
+		name string
+		uri  string
+	}{
+		{
+			name: "when malformed arguments provided, with key only",
+			uri:  "hs:hs_auth:api:v2:provide_access_to_uri?test_arg",
+		},
+		{
+			name: "when malformed arguments provided with key, no value",
+			uri:  "hs:hs_auth:api:v2:provide_access_to_uri?test=",
```

Comment:
I don't know if we should be throwing an error here. This just means that `test` is an empty string, right?

Reply:
Yeah, I need to move that test into the correct section. I just put it here since I wasn't sure if it was a valid URI.